libafl_qemu/src/emu.rs

```diff
@@ -1534,215 +1235,423 @@ impl Emulator {
+    #[must_use]
+    pub fn current_cpu(&self) -> Option<CPU> {
+        self.qemu.current_cpu()
     }
 
-    #[cfg(emulation_mode = "systemmode")]
     #[must_use]
-    pub fn create_fast_snapshot(&self, track: bool) -> FastSnapshot {
-        unsafe {
-            libafl_qemu_sys::syx_snapshot_new(
-                track,
-                true,
-                libafl_qemu_sys::DeviceSnapshotKind_DEVICE_SNAPSHOT_ALL,
-                null_mut(),
-            )
-        }
+    #[allow(clippy::cast_possible_wrap)]
+    pub fn cpu_from_index(&self, index: usize) -> CPU {
+        self.qemu.cpu_from_index(index)
     }
 
-    #[cfg(emulation_mode = "systemmode")]
     #[must_use]
-    pub fn create_fast_snapshot_filter(
-        &self,
-        track: bool,
-        device_filter: &DeviceSnapshotFilter,
-    ) -> FastSnapshot {
-        let mut v = vec![];
-        unsafe {
-            libafl_qemu_sys::syx_snapshot_new(
-                track,
-                true,
-                device_filter.enum_id(),
-                device_filter.devices(&mut v),
-            )
+    pub fn page_from_addr(&self, addr: GuestAddr) -> GuestAddr {
+        self.qemu.page_from_addr(addr)
+    }
+
+    //#[must_use]
+    /*pub fn page_size() -> GuestUsize {
+        unsafe { libafl_page_size }
+    }*/
+
+    pub unsafe fn write_mem(&self, addr: GuestAddr, buf: &[u8]) {
+        self.qemu.write_mem(addr, buf);
+    }
+
+    pub unsafe fn read_mem(&self, addr: GuestAddr, buf: &mut [u8]) {
+        self.qemu.read_mem(addr, buf);
+    }
+
+    #[must_use]
+    pub fn num_regs(&self) -> i32 {
+        self.qemu.num_regs()
+    }
+
+    pub fn write_reg<R, T>(&self, reg: R, val: T) -> Result<(), String>
+    where
+        T: Num + PartialOrd + Copy + Into<GuestReg>,
+        R: Into<i32>,
+    {
+        self.qemu.write_reg(reg, val)
+    }
+
+    pub fn read_reg<R, T>(&self, reg: R) -> Result<T, String>
+    where
+        T: Num + PartialOrd + Copy + From<GuestReg>,
+        R: Into<i32>,
+    {
+        self.qemu.read_reg(reg)
+    }
+
+    pub fn add_breakpoint(&self, mut bp: Breakpoint, enable: bool) {
+        if enable {
+            bp.enable(&self.qemu);
         }
+
+        self.state().breakpoints.borrow_mut().insert(bp);
     }
 
-    #[cfg(emulation_mode = "systemmode")]
-    pub fn restore_fast_snapshot(&self, snapshot: FastSnapshot) {
-        unsafe { libafl_qemu_sys::syx_snapshot_root_restore(snapshot) }
+    pub fn remove_breakpoint(&self, bp: &mut Breakpoint) {
+        bp.disable(&self.qemu);
+
+        self.state().breakpoints.borrow_mut().remove(bp);
     }
 
-    #[cfg(emulation_mode = "systemmode")]
-    pub fn list_devices(&self) -> Vec<String> {
-        let mut r = vec![];
-        unsafe {
-            let devices = libafl_qemu_sys::device_list_all();
-            if devices.is_null() {
-                return r;
-            }
+    pub fn set_breakpoint_addr(&self, addr: GuestAddr) {
```

Comment:
Now we can use `set_breakpoint`, I was using `_addr` to distinguish with the high-level breakpoint functions.
Since we split in 2 structs we can reuse the less confusing notation.